### ISSUE TITLE
[Feature Request] Add option to isolate uncommited branch changes per user

### ISSUE THREAD

**Issue by StoneToken** [treeverse/lakeFS#7277]
Hello!
Currently lakefs has a feature called "uncommited changes" for branch. But two (or more) users can upload and change files in this area. I propose a feature that will allow different users to have separate "uncommited changes" spaces with following requirements:
1. Feature has a toggle to keep it compatible
2. Each user will have his own private uncommited changes area for each branch that he is working
3. Other users won't be able to access private user uncommited changes (unless they are admins?)
4. Uncommited changes will be removed upon branch deletion

I see this feature as a kind of git philosophy -> when using ordinary git a user is feeling safe with his changes unless he explicitly pushes them to remote. For now in lakefs two users may collide in shared branch "uncommited changes" which may be inconvenient.

**Comment by itaiad200** [comment]
Thanks @StoneToken for the suggestion. lakeFS resemblance to git is what makes it UX so intuitive IMO. There are some capabilities that make less sense for object store versioning, like having a local copy. 
In your opinion, what does separate staging area for the same branch allow the user to do, that creating another branch, working on it in isolation and deleting/merging don't?

**Comment by StoneToken** [comment]
Technically you are right - one can accomplish the same goal by means of branches. My story is that when we started playing around with lakefs for our project we stumbled that users first believed that when they pushed their data to branch uncommited changes area they believed that area was private to them, and it was a kind of surprise that it was not. Perhaps this expectation was derived from working on local git repo, but anyway - it feels safe to work with data if you know that your changes won't affect anyone unless commited \ pushed to branch.

**Comment by arielshaqed** [comment]
Hi @StoneToken ,

**tl;dr:** [`lakectl local`](https://lakefs.io/blog/scalable-data-version-control-getting-the-best-of-both-worlds-with-lakefs/) might be a good fit for your needs.

Thanks for opening this issue!  I believe that it goes to the heart of what lakeFS is, so it is not easy in any direction.  Here is my opinion, which perhaps expands on @itaiad200 .

lakeFS needs to be a cross between Git -- for versioning semantics -- and an object store -- for storing objects.  So it supports versions _like Git does_, and it supports objects access _like S3 does_.

But both Git and S3 are more than just versioning and object storage.  Git is also a tool for distributed software development.  A Git repository can push changes to other repositories and pull changes from them.[^1]  Users never interact with a shared repository, every user has their own clone.  One side-effect of working on a clone is that it is typically private.

[^1]: The popular workflow uses GitHub as **the** upstream, and all repositories only go there.

lakeFS cannot go there: ease of use as an object store means some centralization, because data tools such as Spark live on the network.  There is also the matter of the size of data.

One way of re-introducing Git-like local capabilities is provided by [lakeFS local checkouts](https://lakefs.io/blog/scalable-data-version-control-getting-the-best-of-both-worlds-with-lakefs/).  `lakectl local` creates a local copy of a repository commit; you can edit it locally and privately, and upload it back to the repository on demand.

On the server, we intend lakeFS branches to be used for isolation.  As you rightly point out, lakeFS does not support per-branch visibility.  We have no good semantics for this, even adding per-branch ACLs yields surprising results as soon as you add merging.

**Comment by StoneToken** [comment]
Hi, @arielshaqed ! Thanks for feedback, I revised my issue and I think that your proposed way suits better for our needs.

**Comment by arielshaqed** [comment]
> Hi, @arielshaqed ! Thanks for feedback, I revised my issue and I think that your proposed way suits better for our needs.

Cool!  It's always a fine line between solving as many use-cases as possible and not making things too complicated.  I'm glad if in this case we managed both.
